File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/FieldShadowsParentField.cs
using System;

namespace Tests.TestCases
{
    public class Fruit
    {
        protected int ripe;
        protected static int leafs;
    }

    public class Raspberry : Fruit
    {
        private bool ripe;  // Noncompliant {{'ripe' is the name of a field in 'Fruit'.}}
//                   ^^^^
        protected static int leafs; // Compliant, static is ignored
    }
}

// https://github.com/SonarSource/sonar-dotnet/issues/3393
namespace Repro_3393
{
    public class Animal { }
    public class Dog : Animal { }

    public class AnimalContainer
    {
        protected readonly Animal animal;
    }

    public class DogContainer : AnimalContainer
    {
        private new readonly Dog animal; // Compliant, modifier "new" is used to explicitly declare the intention
    }
}

File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/FieldShadowsParentField.vb
Public Class Fruit

    Protected Ripe As Integer
    Protected First, Second As Integer
    Protected Shared Leafs As Integer
    Protected Shadowed As Integer

End Class

Public Class Raspberry
    Inherits Fruit

    Private Ripe As Boolean  ' Noncompliant {{'Ripe' is the name of a field in 'Fruit'.}}
    '       ^^^^
    Protected FirstIsDifferent, Second As Integer  ' Noncompliant {{'Second' is the name of a field in 'Fruit'.}}

    Protected Shared Leafs As Integer       ' Compliant, shared is ignored
    Protected Shadows Shadowed As Integer   ' Compliant, it's explicitly marked as Shadows

End Class
